### PR TITLE
make getrandom behind the wasm feature flag

### PR DESCRIPTION
**Pull request by shekohex**
Greetings, I am attempting to utilize genaric-ec (using cggmp21) by compiling it with the no_std feature and running it in wasm (albeit not in the browser). I initially believed that the compilation would be flawless, however, due to the requirement for getrandom (even without the wasm feature), the compilation fails and complains about the absence of the get random implementation.

Before this change, running `cargo b -p generic-ec --no-default-features --target wasm32-unknown-unknown` result in the following error:

```rs
Compiling getrandom v0.2.12
error: the wasm*-unknown-unknown targets are not supported by default, you may need to enable the "js" feature. For more information see: https://docs.rs/getrandom/#webassembly-support
   --> /home/codespace/.cargo/registry/src/index.crates.io-6f17d22bba15001f/getrandom-0.2.12/src/lib.rs:291:9
    |
291 | /         compile_error!("the wasm*-unknown-unknown targets are not supported by \
292 | |                         default, you may need to enable the \"js\" feature. \
293 | |                         For more information see: \
294 | |                         https://docs.rs/getrandom/#webassembly-support");
    | |________________________________________________________________________^

error[E0433]: failed to resolve: use of undeclared crate or module `imp`
   --> /home/codespace/.cargo/registry/src/index.crates.io-6f17d22bba15001f/getrandom-0.2.12/src/lib.rs:347:9
    |
347 |         imp::getrandom_inner(dest)?;
    |         ^^^ use of undeclared crate or module `imp`

For more information about this error, try `rustc --explain E0433`.
```

With this change, running `cargo b -p generic-ec --no-default-features --target wasm32-unknown-unknown` compiles without issues.

### CODE REVIEW

**Comment by survived** [comment]
You need to enable `wasm` feature by using `cargo build -p generic-ec --target wasm32-unknown-unknown --features wasm,all-curves` command to make it work. We continuously test that it works in [GH actions](https://github.com/dfns/generic-ec/blob/3de295debbe61e2d0f3f4a999372da62968c00e2/.github/workflows/rust.yml#L102C9-L102C93). Or are you saying that enabling `wasm` feature doesn't work?

**Comment by shekohex** [comment]
> You need to enable `wasm` feature by using `cargo build -p generic-ec --target wasm32-unknown-unknown --features wasm,all-curves` command to make it work. We continuously test that it works in [GH actions](https://github.com/dfns/generic-ec/blob/3de295debbe61e2d0f3f4a999372da62968c00e2/.github/workflows/rust.yml#L102C9-L102C93). Or are you saying that enabling `wasm` feature doesn't work?

It works, the only issue is that requires `getrandom/js` and we are working in a non-browser evn, so no js. Hence, removing the need for getrandom all together, since the rng we are using does not depend on getrandom backend.

**Comment by survived** [comment]
Ah I see. In this case, I think that probably `wasm` feature shouldn't be there at all, the client code should either enable `getrandom/js` feature or disable `getrandom`. Could you remove this feature entirely? It'd probably require some other changes to fix CI.

**Comment by survived** [comment]
Also, could you set up commit signing on your side? Repo settings will not allow merging any unsigned/unverified code.

File: generic-ec/src/lib.rs
Comment (survived) [inline generic-ec/src/lib.rs]: Why this change is required?

**Comment by shekohex** [comment]
> Also, could you set up commit signing on your side? Repo settings will not allow merging any unsigned/unverified code.

I do have code signing enabled, github for some reason when using codespaces does not work well. Will sign all the commits and push first.

File: generic-ec/src/lib.rs
Comment (shekohex) [reply generic-ec/src/lib.rs]: If you did not enable serde, it shows unused imports in this module.

Comment (survived) [reply generic-ec/src/lib.rs]: That's strange, but that only means that we need to fix unused imports. `serde` module needs to be present even if feature is disabled

Comment (shekohex) [reply generic-ec/src/lib.rs]: Done.

**Comment by shekohex** [comment]
Hey @survived is it okay to add Nix & flake devenv files to this repo? It makes development consistent across different machines and easier to set up.

**Comment by survived** [comment]
I'll let @maurges decide, he's fan of nix, so he'll probably be reviewing that part :)

**Comment by maurges** [comment]
The nix shell would only contain cargo+rustc (and libc) though? Or do you want to setup a flake for compiling to wasm as well?

**Comment by shekohex** [comment]
Also Related: https://github.com/dfns/stark-curve/issues/4

**Comment by shekohex** [comment]
> The nix shell would only contain cargo+rustc (and libc) though? Or do you want to setup a flake for compiling to wasm as well?

Just the dev shell, no packaging or build using nix are required, so the same tooling and workflow would stay the same, only the initial setup of the repo. Also, I'm using `direnv` so once you `cd dfns/generic-ec` it automatically enters the dev shell.

**Comment by shekohex** [comment]
I've pushed most of the updates, I do have one more test to add in the CI to ensure that WASM + no_std works.

File: generic-ec-curves/src/lib.rs
Comment (maurges) [inline generic-ec-curves/src/lib.rs]: Can't we set no_std unconditionally since we don't use it anywhere?

Comment (shekohex) [reply generic-ec-curves/src/lib.rs]: Yup! will do.

File: rust-toolchain.toml
Comment (survived) [inline rust-toolchain.toml:1]: Why is this file necessary?

Comment (shekohex) [reply rust-toolchain.toml:1]: It ensures that we are using the stable rust, and we use the same components. I must have pushed it during my local setup. Will remove it, if we do not need it in the repo.

**Comment by survived** [comment]
@shekohex actually, we've decided that we don't want to have nix file in the repo. I assume, in your workflow you could put nix file to parent dir and it would work fine with `direnv`?

File: generic-ec-curves/Cargo.toml
Comment (maurges) [inline generic-ec-curves/Cargo.toml]: Style nitpick. Was this an autoformatter? I much preferred the one-line style

Comment (shekohex) [reply generic-ec-curves/Cargo.toml]: Yup, vscode just messed the file, let me fix that.

Comment (survived) [reply generic-ec-curves/Cargo.toml]: Me too. If it's too long, I'd prefer

```toml
[dependencies.elliptic-curve]
version = "0.13"
default-features = false
features = ["sec1", "hash2curve"]
optional = true
```

File: rust-toolchain.toml
Comment (maurges) [reply rust-toolchain.toml:1]: I think it's not necessary. For local development we ourselves want other, different targets, and for CI we select the target by an explicit parameter

**Comment by shekohex** [comment]
> @shekohex actually, we've decided that we don't want to have nix file in the repo. I assume, in your workflow you could put nix file to parent dir and it would work fine with `direnv`?

Yeah, that works, no worries. I just did not push any nix related nor direnv files to the repo, so we are good.

**Comment by shekohex** [comment]
we will also need a new version of stark curve crate so we can get the example that uses no_std to work.

**Comment by survived** [comment]
`stark-curve v0.1.1` is published

**Comment by shekohex** [comment]
Could you please now run the CI?

**Comment by survived** [comment]
I don't see a button to run a CI. Could you push another commit (or do `git commit --amend && git push -f`)? Maybe it'll reappear

**Comment by maurges** [comment]
@shekohex nostd example doesn't build on CI it seems. Do you have access to build logs?

**Comment by shekohex** [comment]
> @shekohex nostd example doesn't build on CI it seems. Do you have access to build logs?

It builds! https://github.com/dfns/generic-ec/actions/runs/7855819173/job/21475955726?pr=17
The only issues are clippy and docs are failing since they try to build it in with std enabled. not sure how we can exclude it from clippy, tests and docs?

**Comment by survived** [comment]
You should be able to exclude them by adding `--exclude` cli flag, e.g. we do that here:
https://github.com/dfns/generic-ec/blob/3de295debbe61e2d0f3f4a999372da62968c00e2/.github/workflows/rust.yml#L61-L62

**Comment by shekohex** [comment]
> You should be able to exclude them by adding `--exclude` cli flag, e.g. we do that here:
> 
> https://github.com/dfns/generic-ec/blob/3de295debbe61e2d0f3f4a999372da62968c00e2/.github/workflows/rust.yml#L61-L62

Thanks for the tip, I've updated the `rust.yml` file, and now it should work, could you approve the CI run?

**Comment by survived** [comment]
You can try the commands locally as well!

**Comment by survived** [comment]
Seems like it requires a workspace flag now

**Comment by shekohex** [comment]
> Seems like it requires a workspace flag now

Done!

**Review by maurges — APPROVED** [review]

**Review by survived — APPROVED** [review]

**Comment by survived** [comment]
Let me just check with our security team what's our procedure, they could also want to review PR

**Review by C0x41lch0x41 — APPROVED** [review]
looks good to me

**Pull request merged**
